### PR TITLE
[#9] As a guest, I can submit the login form

### LINKED ISSUE #9: As a guest, I can submit the login form
## Why

The guest can submit the login form containing `email` and `password`.
So after submitting the server-side will sends some necessary data to the destination endpoint.

## Acceptance Criteria

- Simple validation with valid credentail
- Collect the credentials from the login form

## Resources

- [Sign-in endpoint](https://documenter.getpostman.com/view/11835486/T1LJjTqj#e1562d11-d8a6-4c96-80c2-50d6c4841e42)

### PR DESCRIPTION
**Pull request by Lahphim**
[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=N27Jvfb)

* Resolve [As a guest, I can submit the login form #9 ](https://github.com/Lahphim/react-surveyor-challenge/issues/9)

## What happened 👀

The guest can submit the login form containing `email` and `password`.
So after submitting the server-side will sends some necessary data to the destination endpoint.
 
## Insight 📝

In this PR, the login form can validate the user's credential before sending it to the API server 😇.
For some invalid scenarios, it will show the box of the error messages.
Not just only this but also add some new things here below.

- Add a component to display some error messages
- Add [react-hook-form](https://react-hook-form.com/) for form managment
- Add [yup](https://github.com/jquense/yup) for schema validation
 
## Proof Of Work 📹

| Valid form | Invalid form |
| ---------- | ------------- |
| Screen Shot 2565-06-16 at 23 26 50 | Screen Shot 2565-06-16 at 23 26 13 |

| Preparing the user credential when submission |
| ----------------------------------------------- |
| Screen Shot 2565-06-16 at 23 25 01 |

### CODE REVIEW

File: src/components/FormLogin/index.test.tsx [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
+import userEvent from '@testing-library/user-event';
 
 import { theme } from '@/theme/index';
 
+import { flashToastTestIds } from '../FlashToast/index';
 import FormLogin, { formLoginTestIds } from './index';
 
 describe('FormLogin', () => {
-  it('renders login form', () => {
+  const setup = () =>
```
Comment (hoangmirs) [inline src/components/FormLogin/index.test.tsx:11]: Can we change to use `beforeEach()` instead? 🤔

File: src/components/FormLogin/index.test.tsx [hunk 2]
```diff
@@ -39,4 +44,92 @@ describe('FormLogin', () => {
     expect(labelPassword).toBeVisible();
     expect(inputPassword).toBeVisible();
   });
+
+  describe('given a valid user credentail', () => {
```
Comment (hoangmirs) [inline src/components/FormLogin/index.test.tsx]:
```suggestion
  describe('given a valid user credential', () => {
```

File: src/components/FormLogin/index.test.tsx [hunk 3]
```diff
@@ -39,4 +44,92 @@ describe('FormLogin', () => {
...
+
+      await waitFor(() => {
+        const flashToast = screen.queryByTestId(flashToastTestIds.base);
+
+        expect(flashToast).toBeNull();
+      });
+    });
+  });
+
+  describe('given an INVALID user credentail', () => {
```
Comment (hoangmirs) [inline src/components/FormLogin/index.test.tsx]:
```suggestion
  describe('given an INVALID user credential', () => {
```

File: src/components/FormLogin/index.test.tsx [hunk 4]
```diff
@@ -39,4 +44,92 @@ describe('FormLogin', () => {
...
+      userEvent.click(submit);
+
+      await waitFor(() => {
+        const flashToast = screen.getByTestId(flashToastTestIds.base);
+
+        expect(flashToast).toBeVisible();
+      });
+    });
+
+    describe('given INVALID email', () => {
```
Comment (hoangmirs) [inline src/components/FormLogin/index.test.tsx]:
```suggestion
    describe('given an INVALID email', () => {
```

File: src/components/FormLogin/index.test.tsx [hunk 5]
```diff
@@ -39,4 +44,92 @@ describe('FormLogin', () => {
...
+            flashToastTestIds.list
+          );
+          const { getByText } = within(flashToastMessageList);
+
+          expect(getByText(/email must be a valid email/i)).toBeInTheDocument();
+        });
+      });
+    });
+
+    describe('given INVALID password', () => {
```
Comment (hoangmirs) [inline src/components/FormLogin/index.test.tsx]:
```suggestion
    describe('given an INVALID password', () => {
```

File: src/components/FormLogin/index.test.tsx (hunk 1 above)
Comment (Lahphim) [reply src/components/FormLogin/index.test.tsx:11]: @hoangmirs I try with that before but it didn't pass with the linter.
They suggest doing like this here.

Screen Shot 2565-06-17 at 17 35 53

https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/no-render-in-setup.md

File: src/components/FormLogin/index.test.tsx (hunk 2 above)
Comment (Lahphim) [reply src/components/FormLogin/index.test.tsx]: https://github.com/Lahphim/react-surveyor-challenge/pull/58/commits/db191683c342d4fd70b24fdd670bfe9cb399713d

File: src/components/FormLogin/index.test.tsx (hunk 3 above)
Comment (Lahphim) [reply src/components/FormLogin/index.test.tsx]: https://github.com/Lahphim/react-surveyor-challenge/pull/58/commits/db191683c342d4fd70b24fdd670bfe9cb399713d

File: src/components/FormLogin/index.test.tsx (hunk 5 above)
Comment (Lahphim) [reply src/components/FormLogin/index.test.tsx]: https://github.com/Lahphim/react-surveyor-challenge/pull/58/commits/644e7c5b7ba3a0315d8707c45a932adc7a28e3b5

File: src/components/FormLogin/index.test.tsx (hunk 4 above)
Comment (Lahphim) [reply src/components/FormLogin/index.test.tsx]: https://github.com/Lahphim/react-surveyor-challenge/pull/58/commits/644e7c5b7ba3a0315d8707c45a932adc7a28e3b5

**Review by hoangmirs — APPROVED** [review]

File: src/components/FlashToast/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      const flashToastMessageList = screen.getByTestId(flashToastTestIds.list);
+      const firstMessage = within(flashToastMessageList).getByText(
+        validProps.messageList[0]
+      );
+      const secondMessage = within(flashToastMessageList).getByText(
+        validProps.messageList[1]
+      );
+
+      expect(firstMessage).toBeInTheDocument();
+      expect(secondMessage).toBeInTheDocument();
```
Comment (carryall) [inline src/components/FlashToast/index.test.tsx]: How about using `.toBeVisible()` here? 💭 

```suggestion
      expect(firstMessage).toBeVisible();
      expect(secondMessage).toBeVisible();
```

File: src/components/FormLogin/index.tsx [hunk 7]
```diff
@@ -15,42 +21,83 @@ export const formLoginTestIds = {
...
+      />
+
+      <form onSubmit={handleSubmit(onSubmit)}>
+        <SimpleGrid mt="8" spacing="6" data-test-id={formLoginTestIds.base}>
+          <FormControl>
+            <FormLabel
+              htmlFor="email"
+              data-test-id={formLoginTestIds.labelEmail}
+            >
+              Email
```
Comment (carryall) [inline src/components/FormLogin/index.tsx:67]: Just a small question here, do you have any plan to work on localization?

File: src/pages/index.page.tsx [hunk 8]
```diff
@@ -141,6 +142,21 @@ const Home = () => {
...
+            <h2>:: Flash Toast</h2>
+            <div>
+              <FlashToast
+                title="Error"
+                messageList={[
+                  'lorem ipsum dolor sit amet',
+                  'consectetur adipiscing elit',
+                  'integer molestie lorem at massa',
+                  'facilisis in pretium nisl aliquet',
+                ]}
```
Comment (carryall) [inline src/pages/index.page.tsx:156]: I am just wondering do we need to set the dummy string here? 🤔

**Review by carryall — APPROVED** [review]
Only some small comments, the rest LGTM

File: src/pages/index.page.tsx (hunk 8 above)
Comment (Lahphim) [reply src/pages/index.page.tsx:156]: On this page, I use it as a styleguide page.
it's just some require text for rendering the component.

File: src/components/FormLogin/index.tsx (hunk 7 above)
Comment (Lahphim) [reply src/components/FormLogin/index.tsx:67]: @carryall No plan for that 🖖

Comment (Lahphim) [reply src/components/FormLogin/index.tsx:67]: No plan for that 🙌

File: src/components/FlashToast/index.test.tsx (hunk 6 above)
Comment (Lahphim) [reply src/components/FlashToast/index.test.tsx]: https://github.com/Lahphim/react-surveyor-challenge/pull/58/commits/fda34832a965c0886b44cf11b0ff88c68938e99a

**Pull request merged**